spec.bs

```diff
@@ -491,12 +494,243 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
+
+A [=tokenizer=] has an associated <dfn for=tokenizer>token list</dfn>, a [=/token list=], initially an empty [=list=].
+
+A [=tokenizer=] has an associated <dfn for=tokenizer>index</dfn>, a number, initially 0.
+
+A [=tokenizer=] has an associated <dfn for=tokenizer>next index</dfn>, a number, initially 0.
+
+A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a unicode code point, initially null.
+
+<div algorithm>
+  To <dfn>get the next code point</dfn> for a given [=tokenizer=] |tokenizer|:
+
+  1. Set |tokenizer|'s [=tokenizer/code point=] to the unicode code point in |tokenizer|'s [=tokenizer/input=] at the position indicated by |tokenizer|'s [=tokenizer/next index=].
+  1. Increment |tokenizer|'s [=tokenizer/next index=] by 1.
+</div>
+
+<div algorithm>
+  To <dfn>seek and get the next code point</dfn> for a given [=tokenizer=] |tokenizer| and number |index|:
+
+  1. Set |tokenizer|'s [=tokenizer/next index=] to |index|.
+  1. Run [=get the next code point=] given |tokenizer|.
+</div>
+
+<div algorithm>
+  To <dfn>add a token</dfn> for a given [=tokenizer=] |tokenizer|, [=token/type=] |type|, number |next position|, number |value position|, and number |value length|:
+
+  1. Let |token| be a new [=token=].
+  1. Set |token|'s [=token/type=] to |type|.
+  1. Set |token|'s [=token/index=] to |tokenizer|'s [=tokenizer/index=].
+  1. Set |token|'s [=token/value=] to the substring within |tokenizer|'s [=tokenizer/input=] starting at |value position| with |value length|.
+  1. Set |tokenizer|'s [=token/index=] to |next position|.
+</div>
+
+<div algorthm>
+  To <dfn>add a token with default length</dfn> for a given [=tokenizer=] |tokenizer|, [=token/type=] |type|, number |next position|, and number |value position|.
+
+  1. Let |computed length| be |next position| &minus; |value position|.
+  1. Run [=add a token=] given |tokenizer|, |type|, |next position|, |value position|, and |computed length|.
+</div>
+
+<div algorithm>
+  To <dfn>add a token with default position and length</dfn> for a given [=tokenizer=] |tokenizer| and [=token/type=] |type|:
+
+  1. Run [=add a token with default length=] given |tokenizer|, |type|, |tokenizer|'s [=tokenizer/next index=], and |tokenizer|'s [=tokenizer/index=].
+</div>
+
+<div algorithm>
+  To <dfn>process a tokenizing error</dfn> for a given [=tokenizer=] |tokenizer|, a number |next position|, and a number |value position|:
+
+  1. If |tokenizer|'s [=tokenizer/policy=] is "[=tokenize policy/strict=]", then throw a {{TypeError}}.
+  1. [=Assert=]: |tokenizer|'s [=tokenizer/policy=] is "[=tokenize policy/lenient=]".
+  1. Run [=add a token with default length=] given |tokenizer|, "[=token/type/invalid-char=]", |next position|, and |value position|.
+</div>
+
+<div algorithm>
+  To <dfn>process a tokenizing error with default position</dfn> for a given [=tokenizer=] |tokenizer|:
```

Comment:
This and "is a valid name code point" only have one call site. It might be clearer to inline them, or it might be nice to keep them as semantic helpers.

Reply:
I personally like having them separate to help keep the main algorithm as clean as possible.  Maybe we could inline the one about default position.  I thought I used that more.

Reply:
Ok.  I inlined the "process a tokenizing error with default position" call.  Keeping the "is a valid name code point" because I think the algorithm name is informative of the purpose.